core/src/light_protocol/query_handler.rs

```diff
@@ -0,0 +1,458 @@
+// Copyright 2019 Conflux Foundation. All rights reserved.
+// Conflux is free software and distributed under GNU General Public License.
+// See http://www.gnu.org/licenses/
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    sync::Arc,
+};
+
+extern crate futures;
+use futures::{
+    sync::oneshot::{self, Sender},
+    Async, Future,
+};
+
+use io::TimerToken;
+use parking_lot::RwLock;
+use priority_send_queue::SendQueuePriority;
+use rand::Rng;
+use rlp::Rlp;
+
+use cfx_bytes::Bytes;
+use cfx_types::H256;
+use primitives::{BlockHeader, EpochNumber, StateRoot};
+
+use crate::{
+    consensus::{ConsensusGraph, DEFERRED_STATE_EPOCH_COUNT},
+    network::{
+        throttling::THROTTLING_SERVICE, Error as NetworkError, HandlerWorkType,
+        NetworkContext, NetworkProtocolHandler, NetworkService, PeerId,
+        ProtocolId,
+    },
+    statedb::StateDb,
+    storage::{
+        state_manager::StateManagerTrait, SnapshotAndEpochIdRef, StateProof,
+    },
+};
+
+use super::{
+    message::{
+        GetStateEntry, GetStateRoot, Message, MsgId, RequestId,
+        StateEntry as GetStateEntryResponse, StateRoot as GetStateRootResponse,
+    },
+    Error, ErrorKind,
+};
+
+// TODO(thegaram): remove this once the `network.with_context`
+// call is moved to QueryService
+const LIGHT_PROTOCOL_ID: ProtocolId = *b"clp"; // Conflux Light Protocol
+
+const POLL_PERIOD_MS: u64 = 100;
+const MAX_POLL_TIME_MS: u64 = 1000;
+
+#[derive(Debug)]
+pub(super) enum QueryResult {
+    StateEntry(Option<Vec<u8>>),
+    StateRoot(StateRoot),
+}
+
+struct PendingRequest {
+    msg: Box<dyn Message>,
+    sender: Sender<QueryResult>,
```

Comment:
Is it possible to handle the request with Sender more generic, e.g. define the sender as Sender<Response>?

Reply:
Not sure I understand what you mean here, could you elaborate?

Reply:
Sorry for typo. I mean make the sender more generic, e.g. Sender<Response>